### PR TITLE
Changes to abstract compilation services from C# invoker

### LINKED ISSUE #218: Compilation service unit tests
Make the required "compiler service" changes to enabled better tests and add compilation unit tests.

### PR DESCRIPTION
**Pull request by fabiocav**
#### Summary of changes
The work in this PR is primarily code cleanup and refactoring to abstract the compilation logic (introducing an `ICompilationService` and related interfaces) from the invoker, where that C# specific logic was previously contained.
This resolves #218 , provides an easier/better model to plug other .NET languages in and supports some of the changes we're currently working on for the invoker updates.

### CODE REVIEW

**Comment by fabiocav** [comment]
@dsyme, these are the changes I've shared with you. I tried to incorporate things you guys needed where there were clear gaps, but please let me know if you feel additional changes are required or have any other concerns.

**Comment by dsyme** [comment]
This looks good.  If you merge it I can reconcile it with the F# changes

File: src/WebJobs.Script/Description/DotNet/DotNetCompilationServiceFactory.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        {
+            get
+            {
+                return SupportedScriptTypes;
+            }
+        }
+
+        public ICompilationService CreateService(ScriptType scriptType, IFunctionMetadataResolver metadataResolver)
+        {
+            switch (scriptType)
```
Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/DotNetCompilationServiceFactory.cs]: Will this ever be something other than CSharp? If not, then no switch needed?

File: src/WebJobs.Script/Description/DotNet/DotNetCompilationServiceFactory.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            {
+                case ScriptType.CSharp:
```
Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/DotNetCompilationServiceFactory.cs]: So the plan is to slot FSharp in here as well?

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs [hunk 3]
```diff
@@ -10,63 +10,55 @@
...
+        private readonly ICompilationService _compilationService;
 
-        private CSharpFunctionSignature _functionSignature;
+        private FunctionSignature _functionSignature;
         private IFunctionMetadataResolver _metadataResolver;
         private Action _reloadScript;
         private Action _restorePackages;
         private Action<MethodInfo, object[], object[], object> _resultProcessor;
         private FunctionValueLoader _functionValueLoader;
+        private string[] watchedFileTypes;
```
Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: nit: follow the _ pattern for fields :)

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs [hunk 4]
```diff
@@ -367,7 +364,7 @@ private ImmutableArray<Diagnostic> GetFunctionDiagnostics(Diagnostic diagnostic)
                         "The reference '{0}' is part of the referenced NuGet package '{1}'. Package assemblies are automatically referenced by your Function and do not require a '#r' directive.",
```
Comment (mathewc) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: Does FSharp use the same #r package syntax?

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 5]
```diff
@@ -206,7 +206,7 @@ protected virtual void Initialize()
             {
                 new ScriptFunctionDescriptorProvider(this, ScriptConfig),
                 new NodeFunctionDescriptorProvider(this, ScriptConfig),
-                new CSharpFunctionDescriptionProvider(this, ScriptConfig)
+                new DotNetFunctionDescriptionProvider(this, ScriptConfig, new DotNetCompilationServiceFactory())
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Can the compiler service param be optional?

**Comment by mathewc** [comment]
:ship:

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs (hunk 4 above)
Comment (ahmelsayed) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: `fsx` does. Pretty sure that's where Roslyn got the syntax from

File: src/WebJobs.Script/Description/DotNet/DotNetCompilationServiceFactory.cs (hunk 2 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/DotNet/DotNetCompilationServiceFactory.cs]: That's the idea.

File: src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs (hunk 3 above)
Comment (fabiocav) [inline src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs]: Good catch! I thought CA would yell at me for that :S

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 5 above)
Comment (fabiocav) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Sure. I'll make this change.

**Comment by fabiocav** [comment]
@dsyme this has been merged into `dev`

**Comment by dsyme** [comment]
Should I retarget [this PR](https://github.com/Azure/azure-webjobs-sdk-script/pull/279#issuecomment-217535909) to dev?

**Comment by mathewc** [comment]
Yes, our default branch is now dev. master is now our release branch. Sorry if this transition causes you troubles :)

**Comment by dsyme** [comment]
np, will do

**Pull request merged**